src/components/DependencyTree/DependencyTree.tsx

```diff
@@ -0,0 +1,262 @@
+      if (serviceData!.build) {
+        return buildService.getDependencyGraph({ id: serviceData!.build!.id }, requestConfig);
+      } else {
+        return groupBuildService.getDependencyGraph({ id: serviceData!.groupBuild!.id }, requestConfig);
+      }
+    }, [])
+  );
+
+  const refreshDataContainer = dataContainer.refresh;
+
+  const refreshComponent = () => {
+    setVisible(false);
+    setTimeout(() => {
+      setVisible(true);
+    }, 0);
+  };
+
+  const onClickExpandLevel1 = () => {
+    setAllExpanded(undefined);
+    setDefaultExpandByLevel(1, dependencyStructure!);
+  };
+  const onClickExpandLevel2 = () => {
+    setAllExpanded(undefined);
+    setDefaultExpandByLevel(2, dependencyStructure!);
+  };
+  const onClickExpandAll = () => {
+    setAllExpanded(allExpanded !== undefined ? !allExpanded : true);
+  };
+
+  const onClickExpandAllFailed = () => {
+    setAllExpanded(undefined);
+    setExpandFailed(dependencyStructure!);
+  };
+
+  const setExpandFailed = (dependStructure: IDependencyDataItem) => {
+    dependStructure.defaultExpanded = [
+      'FAILED',
+      'REJECTED',
+      'REJECTED_FAILED_DEPENDENCIES',
+      'SYSTEM_ERROR',
+      'ENQUEUED',
+      'WAITING_FOR_DEPENDENCIES',
+      'BUILDING',
+    ].includes(dependStructure.data.status!);
+    dependStructure.children?.forEach((child) => setExpandFailed(child));
+    if (dependStructure.level === 0) {
+      refreshComponent();
+    }
+  };
+
+  const setDefaultExpandByLevel = (level: number, dependStructure: IDependencyDataItem) => {
+    dependStructure.defaultExpanded = dependStructure.level < level;
+    dependStructure.children?.forEach((child) => setDefaultExpandByLevel(level, child));
+    if (dependStructure.level === 0) {
+      refreshComponent();
+    }
+  };
+
+  const generateTreeItem = (build: IDependencyBuild) =>
+    build && (
+      <span className={styles['tree-item']}>
+        <BuildStatus build={build} includeBuildLink includeConfigLink long />
+      </span>
+    );
+
+  useEffect(() => {
+    setBuildItem(build ? build : groupBuild);
+    refreshDataContainer({ serviceData: { build: build, groupBuild: groupBuild } });
+  }, [build, groupBuild, refreshDataContainer]);
+
+  useEffect(() => {
+    const getRootNodes = (edgesData: Array<IGraphEdge>, nodesData: Map<string, IDependencyBuild>) => {
+      const rootNodes: Array<string> = [];
+      nodesData.forEach((node) => {
+        !edgesData.map((edge) => edge.target).includes(node.id) && rootNodes.push(node.id);
+      });
+      return rootNodes;
+    };
+
+    const attachChildFromEdges = (
+      currentNode: IDependencyDataItem,
+      edgesData: Array<IGraphEdge>,
+      nodesData: Map<string, IDependencyBuild>,
+      level: number
+    ) => {
+      level = level ? level : 1;
+      if (level === 1 && !isBuild(buildItem)) {
+        const rootNodes: Array<string> = level === 1 ? getRootNodes(edgesData, nodesData) : [];
```

Comment:
The level can be other than 1. this is a recursive function so it will pass level+1 for every child of it. see https://github.com/DnsZhou/pnc-web-ui-react/blob/d3e71c56fc9b00c291355ff8c5e0ed2577543394/src/components/DependencyTree/DependencyTree.tsx#L186

Reply:
> no point using ternary operator - level will always be 1, since it is required by if condition

The level can be other than 1. this is a recursive function so it will pass level+1 for every child of it. see https://github.com/DnsZhou/pnc-web-ui-react/blob/d3e71c56fc9b00c291355ff8c5e0ed2577543394/src/components/DependencyTree/DependencyTree.tsx#L186

Reply:
Yes, sure, but inside if, the level will always be equal to one, since it is condition required by if.

Reply:
aha got it, missed that part